server/app/services/cloud/websocket_client.rb

```diff
@@ -91,91 +68,164 @@ def connecting?
-      info "cloud connection opened to #{self.api_uri}"
-      subscribe_events(EventStream.channel)
+    def on_open
       @connected = true
       @connecting = false
-    end
 
-    # @param [Faye::WebSocket::API::Event] event
-    def on_message(event)
-      debug "Received websocket message"
-      if leader?
-        data = MessagePack.unpack(event.data.pack('c*'))
-        EM.defer {
-          if request_message?(data)
-            debug "Creating RPC request"
-            response = rpc_server.handle_request(data)
-            send_message(MessagePack.dump(response).bytes)
-          elsif notification_message?(data)
-            rpc_server.handle_notification(data)
-          end
-        }
+      begin
+        ssl_cert = @ws.ssl_cert!
+      rescue Kontena::Websocket::SSLVerifyError => ssl_error
+        ssl_cert = ssl_error.cert
+      else
+        ssl_error = nil
+      end
+
+      ssl_verify = @ws.ssl_verify?
+
+      if ssl_cert && ssl_error
+        warn "Connected to #{@ws.url} with ssl errors: #{ssl_error} (subject #{ssl_cert.subject}, issuer #{ssl_cert.issuer})"
+      elsif ssl_error
+        warn "Connected to #{@ws.url} with ssl errors: #{ssl_error}"
+      elsif ssl_cert && !ssl_verify
+        warn "Connected to #{@ws.url} without ssl verify: #{ssl_cert.subject} (issuer #{ssl_cert.issuer})"
+      elsif ssl_cert
+        info "Connected to #{@ws.url} with ssl verify: #{ssl_cert.subject} (issuer #{ssl_cert.issuer})"
       else
+        info "Connected to #{@ws.url} without ssl"
+      end
+
+      subscribe_events(EventStream.channel)
+    end
+
+    # @param [String, Array] msg
+    def on_message(msg)
+      unless leader?
         debug "Ignoring request because not leader"
+        return
+      end
+
+      data = MessagePack.unpack(msg.pack('c*'))
+
+      if request_message?(data)
+        debug "RPC request #{data[2]}"
+        response = rpc_server.handle_request(data)
+        send_message(MessagePack.dump(response).bytes)
+      elsif notification_message?(data)
+        debug "RPC notification #{data[2]}"
+        rpc_server.handle_notification(data)
+      else
+        warn "Unknown RPC: #{data}"
       end
     rescue => exc
-      error exc.message
+      error exc
     end
 
-    # @param [Faye::WebSocket::API::Event] event
-    def on_close(event)
-      @ping_timer = nil
+    # @param code [Integer]
+    # @param reason [String]
+    def on_close(code, reason)
       @connected = false
       @connecting = false
-      @ws = nil
-      if event.code == 1002
-        handle_invalid_token
+
+      case code
+      when 1002
+        error 'cloud does not accept our access token'
+      else
+        info "cloud connection closed with code #{code}: #{reason}"
       end
-      info "cloud connection closed with code: #{event.code}"
+
       unsubscribe_events
+    end
+
+     # @return [Kontena::Websocket::Client]
+    def ws
+      @ws.tap do |ws|
+        fail "not connected" unless ws
```

Comment:
The only place this `self.ws` is actually used is in the `send_message`, so that calls from the separate `MongoPubsub.subscribe` thread will fail with something other than `NoMethodError: undefined method 'send' for nil` after the `@ws` gets reset to `nil`.

The `ws.send` itself can already fail after a websocket close.